File: src/main/java/controllers/supply/SuppliersViewController.java
package controllers.supply;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.stage.Modality;
import javafx.stage.Stage;
import models.Supplier;
import network.Api;
import network.RetrofitClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import utils.Dialogs;

import java.io.IOException;
import java.util.List;

/**
 * Created by andrey on 25.07.17.
 */
public class SuppliersViewController {
    Api api = RetrofitClient.INSTANCE.getApiService();
    private ObservableList<Supplier> suppliersOL = FXCollections.observableArrayList();

    @FXML
    private TableView<Supplier> suppliersTable;
    @FXML
    private TableColumn<Supplier, String> nameColumn;
    @FXML
    private TableColumn<Supplier, String> addressColumn;


    private void loadSuppliersData() {
        Call<List<Supplier>> call = api.suppliers();
        call.enqueue(new Callback<List<Supplier>>() {
            @Override
            public void onResponse(Call<List<Supplier>> call, Response<List<Supplier>> response) {
                if (response.code() == 200) {
                    suppliersOL = FXCollections.observableArrayList(response.body());
                    suppliersTable.setItems(suppliersOL);
                } else {
                    Dialogs.INSTANCE.showErrorDialog("Ошибка запроса на сервер!");
                }
            }

            @Override
            public void onFailure(Call<List<Supplier>> call, Throwable t) {
                Dialogs.INSTANCE.showExeptionDialog(t.getMessage());
            }
        });
    }

    @FXML
    private void initialize() {
        // Инициализация таблицы адресатов с двумя столбцами.
        nameColumn.setCellValueFactory(new PropertyValueFactory<Supplier, String>("name"));
        addressColumn.setCellValueFactory(new PropertyValueFactory<Supplier, String>("urAddress"));
        loadSuppliersData();


    }


    public void handleAdd(ActionEvent actionEvent) {
        try {
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("/view/supply/SuppliersEditView.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Новый поставщик");
            stage.setMinHeight(150);
            stage.setMinWidth(400);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);
            stage.initOwner(((Node) actionEvent.getSource()).getScene().getWindow());

            SupplierEditController controller = loader.getController();

            stage.showAndWait();

            if (controller.isOkClicked()) {
                loadSuppliersData();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void handleEdit(Supplier supplier) {
        try {
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("/view/supply/SuppliersEditView.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Редактировать поставщика");
            stage.setMinHeight(150);
            stage.setMinWidth(400);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);

            SupplierEditController controller = loader.getController();
            controller.setSupplier(supplier);

            stage.showAndWait();

            if (controller.isOkClicked()) {
                loadSuppliersData();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void handleDel(Supplier supplier) {
        try {
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("/view/supply/DelSupplier.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Удалить поставщика");
            stage.setMinHeight(150);
            stage.setMinWidth(400);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);

            SupplierEditController controller = loader.getController();
            controller.setSupplier(supplier);

            stage.showAndWait();

            if (controller.isOkClicked()) {
                loadSuppliersData();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: src/main/kotlin/controllers/supply/SuppliersViewController.kt
package controllers.supply

import controllers.LoadController
import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.control.cell.PropertyValueFactory
import javafx.stage.Modality
import kotlinx.coroutines.experimental.javafx.JavaFx
import kotlinx.coroutines.experimental.launch
import models.supply.Supplier
import utils.CloseListener
import utils.Dialogs


/**
 * Created by andrey on 25.07.17.
 */
class SuppliersViewController : LoadController<Boolean>() {

    @FXML private lateinit var suppliersTable: TableView<Supplier>

    @FXML private lateinit var nameColumn: TableColumn<Supplier, String>

    @FXML private lateinit var name: Label
    @FXML private lateinit var inn: Label
    @FXML private lateinit var type: Label
    @FXML private lateinit var urAddress: Label
    @FXML private lateinit var physAddress: Label
    @FXML private lateinit var requisites: Label
    @FXML private lateinit var name1: Label
    @FXML private lateinit var phone1: Label
    @FXML private lateinit var name2: Label
    @FXML private lateinit var phone2: Label

    @FXML
    private fun initialize() {
        nameColumn.setCellValueFactory(PropertyValueFactory("name"))
        suppliersTable.selectionModel.selectedItemProperty().addListener { _, _, newValue -> newValue?.let { showSupplier(newValue) } }

        loadSuppliersData()

        val contextMenu = ContextMenu()
        val edit = MenuItem("Редактировать")
        edit.onAction = EventHandler {
            val item = suppliersTable.selectionModel.selectedItem
            editSupplier(item)
        }
        contextMenu.items.setAll(edit)
        suppliersTable.contextMenu = contextMenu
    }

    fun handleAdd(actionEvent: ActionEvent) {
        SupplierEditController.show(owner = actionEvent.source as Node) { result ->
            if (result) {
                loadSuppliersData()
            }
        }
    }

    fun editSupplier(supplier: Supplier) {
        SupplierEditController.show(owner = name as Node, supplier = supplier) { result ->
            if (result) {
                loadSuppliersData()
            }
        }
    }

    fun showSupplier(supplier: Supplier) {
        name.text = supplier.name
        inn.text = supplier.inn
        // нужен метод который преводит тип в строку
        //type.text = supplier.type
        urAddress.text = supplier.urAddress
        physAddress.text = supplier.physAddress
        requisites.text = supplier.requisites
        name1.text = supplier.name1
        phone1.text = supplier.phone1
        name2.text = supplier.name2
        phone2.text = supplier.phone2
    }

    private fun loadSuppliersData() = launch(JavaFx) {
        val result = api.suppliers().await()
        if (result.isSuccessful()) {
            val suppliers = result.notNullResult()
            suppliersTable.items = FXCollections.observableArrayList(suppliers)
        } else {
            Dialogs.showExeptionDialog(result.error)
        }
    }

    companion object {
        fun show(owner: Node, callback: CloseListener<Boolean>) {
            LoadController.show(owner, callback,
                    path = "/view/supply/SuppliersView.fxml",
                    title = "Поставщики",
                    isResizable = false,
                    modality = Modality.WINDOW_MODAL)
        }
    }
}
